### PR TITLE
Add BlockBreakProgressUpdateEvent

### PR DESCRIPTION
**Pull request by Badbird5907**
Added a `BlockBreakProgressUpdateEvent`, which allows you to listen for block break animation progress updates

### CODE REVIEW

File: patches/server/1054-add-BlockBreakProgressUpdateEvent.patch [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
++++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
+@@ -1772,6 +1772,12 @@ public class ServerLevel extends Level implements WorldGenLevel {
+         if (entity instanceof Player) entityhuman = (Player) entity;
+         // CraftBukkit end
+ 
++        // Paper start - Add BlockBreakProgressUpdateEvent
++        org.bukkit.craftbukkit.block.CraftBlock bukkitBlock = org.bukkit.craftbukkit.block.CraftBlock.at(this, pos);
++        io.papermc.paper.event.block.BlockBreakProgressUpdateEvent event = new io.papermc.paper.event.block.BlockBreakProgressUpdateEvent(bukkitBlock, progress, entity.getBukkitEntity());
++        Bukkit.getPluginManager().callEvent(event);
```
Comment (yannicklamprecht) [inline patches/server/1054-add-BlockBreakProgressUpdateEvent.patch]: Instead use Event#call().

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

**Review by electronicboy — APPROVED** [review]
Some trivial cleanups would be ideal, i.e. the event can be inlined with callEvent() as noted, maybe dubious on the progress thing technically being magical but, that is already exposed within the API to some degree

**Comment by Badbird5907** [comment]
updated to use Event#callEvent

File: patches/server/1054-add-BlockBreakProgressUpdateEvent.patch [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
++++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
+@@ -1772,6 +1772,12 @@ public class ServerLevel extends Level implements WorldGenLevel {
+         if (entity instanceof Player) entityhuman = (Player) entity;
+         // CraftBukkit end
+ 
++        // Paper start - Add BlockBreakProgressUpdateEvent
++        org.bukkit.craftbukkit.block.CraftBlock bukkitBlock = org.bukkit.craftbukkit.block.CraftBlock.at(this, pos);
++        io.papermc.paper.event.block.BlockBreakProgressUpdateEvent event = new io.papermc.paper.event.block.BlockBreakProgressUpdateEvent(bukkitBlock, progress, entity.getBukkitEntity());
++        event.callEvent();
```
Comment (yannicklamprecht) [inline patches/server/1054-add-BlockBreakProgressUpdateEvent.patch]:
```suggestion
+        // Paper start - Add BlockBreakProgressUpdateEvent
+        org.bukkit.craftbukkit.block.CraftBlock bukkitBlock = org.bukkit.craftbukkit.block.CraftBlock.at(this, pos);
+        io.papermc.paper.event.block.BlockBreakProgressUpdateEvent event = new io.papermc.paper.event.block.BlockBreakProgressUpdateEvent(bukkitBlock, progress, entity.getBukkitEntity()).callEvent();
```

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

**Review by yannicklamprecht — APPROVED** [review]

**Review by lynxplay — APPROVED** [review]

File: patches/api/0468-add-BlockBreakProgressUpdateEvent.patch [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
++    @ApiStatus.Internal
++    public BlockBreakProgressUpdateEvent(@NotNull final Block block, final float progress, @NotNull final Entity entity) {
++        super(block);
++        this.progress = progress;
++        this.entity = entity;
++    }
++
++    /**
++     * The progress of the block break the progress from 0.0 - 1.0 where 0 is no damage and
++     * 1.0 is the most damaged
```
Comment (NoahvdAa) [inline patches/api/0468-add-BlockBreakProgressUpdateEvent.patch]: This sentence is kinda weird (copy paste mistake?). Maybe reformat to something like
```suggestion
+     * The progress of the block break
+     * <p>
+     * The progress ranges from 0.0 - 1.0, where 0 is no damage and
+     * 1.0 is the most damaged
```

Comment (Badbird5907) [reply patches/api/0468-add-BlockBreakProgressUpdateEvent.patch]: Whoops yeah, will do

**Review by lynxplay — APPROVED** [review]

**Pull request merged**